src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -109,49 +81,47 @@ internal ReturnTypeAndDiagnostics(TypeSymbol returnType, ImmutableArray<Diagnost
         /// </summary>
         internal Binder ScopeBinder { get; }
 
+        public Binder ParameterBinder => _binder;
+
         internal void GetDeclarationDiagnostics(DiagnosticBag addTo)
         {
-            // Force attribute binding for diagnostics
-            foreach (var typeParam in _typeParameters)
+            lock (_lazyLock)
```

Comment:
Yes -- without the lock there's no synchronization on adding the diagnostics to the bag. You could end up with two threads calling GetDeclarationDiagnostics and them each adding duplicate declaration diagnostics to the bag.